src/NUnitEngine/nunit.engine/Services/AgentStore.cs

```diff
@@ -10,45 +10,57 @@ namespace NUnit.Engine.Services
     /// <summary>
     /// Defines the agent tracking operations that must be done atomically.
     /// </summary>
-    internal sealed partial class AgentStore
+    internal sealed class AgentStore
     {
-        private readonly Dictionary<Guid, AgentRecord> _agentsById = new Dictionary<Guid, AgentRecord>();
+        private static readonly Logger log = InternalTrace.GetLogger(typeof(AgentStore));
+
+        private readonly object LOCK = new object();
+
+        private readonly Dictionary<Guid, AgentRecord> _agentIndex = new Dictionary<Guid, AgentRecord>();
+        private readonly Dictionary<Process, AgentRecord> _processIndex = new Dictionary<Process, AgentRecord>();
```

Comment:
You shouldn't need the dictionary on process. Process termination should be handled by the `LocalProcessAgent`